lib/rules/default-param-last.js

```diff
@@ -5,6 +5,35 @@
 
 "use strict";
 
+/**
+ * checks if node is optional parameter
+ * @param {ASTNode} node the node to be evaluated
+ * @returns {boolean} whether the node is optional
+ */
+function isOptionalParam(node) {
+    return (
+        (node.type === "ArrayPattern" ||
+            node.type === "AssignmentPattern" ||
+            node.type === "Identifier" ||
+            node.type === "ObjectPattern" ||
+            node.type === "RestElement") &&
+        node.optional
```

Comment:
I'm curious about this logic. Why isn't it enough to just check `node.optional`?

If we do need to check for all of these node types, then it's more efficient to place them all in a `Set` and use `has()`.

Reply:
Ha, yeah, this is a code style thing. In the typescript-eslint repo, since `TSESTree.Node` is strictly typed as a union of nodes, you can't access arbitrary properties that only exist on some nodes. We're very cautious with arbitrary property accesses and require type narrowing: e.g. https://github.com/typescript-eslint/typescript-eslint/issues/8226.

Since we don't do that here, I simplified to just check `node.optional` inline.